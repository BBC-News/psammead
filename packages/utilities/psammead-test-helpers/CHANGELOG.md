# Psammead Test Helpers Changelog

| Version | Description |
|---------|-------------|
| 0.3.0   | [PR#224](https://github.com/BBC-News/psammead/pull/224) Adds a helper method for testing package export values and uses it to test its own exports. |
| 0.2.1   | [PR#255](https://github.com/BBC-News/psammead/pull/255) Bumps package version to fix what appears to be an erroneously-published version. |
| 0.1.2   | [PR#212](https://github.com/BBC-News/psammead/pull/212) Update package description and README. |
| 0.1.1   | [PR#173](https://github.com/BBC-News/psammead/pull/173) Update PRs welcome link |
| 0.1.0   | [PR#80](https://github.com/BBC-News/psammead/pull/80) Create initial package, pulled in from [simorgh](https://github.com/BBC-News/simorgh). |
