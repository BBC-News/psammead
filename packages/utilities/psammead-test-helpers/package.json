{
  "name": "@bbc/psammead-test-helpers",
  "version": "0.3.0",
  "main": "dist/index.js",
  "description": "A collection of helper methods for implementing Jest snapshot tests for styled-components, required by many Psammead components.",
  "repository": {
    "type": "git",
    "url": "https://github.com/BBC-News/psammead/tree/latest/packages/utilities/psammead-test-helpers"
  },
  "author": {
    "name": "Psammead Maintainers",
    "email": "PsammeadMaintainers@bbc.co.uk"
  },
  "license": "Apache-2.0",
  "bugs": {
    "url": "https://github.com/bbc/psammead/issues"
  },
  "homepage": "https://github.com/BBC-News/psammead/blob/latest/packages/utilities/psammead-test-helpers/README.md",
  "keywords": [
    "bbc",
    "utilities",
    "test",
    "helpers",
    "jest",
    "snapshot"
  ],
  "dependencies": {
    "jest-styled-components": "^6.3.1",
    "react-test-renderer": "^16.6.3"
  }
}
