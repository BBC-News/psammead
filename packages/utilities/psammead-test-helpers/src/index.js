import renderer from 'react-test-renderer';
import 'jest-styled-components';
import ShallowRenderer from 'react-test-renderer/shallow';

export const shouldMatchSnapshot = (title, component) => {
  it(title, () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });
};

export const shallowRender = component => {
  const shallowRenderer = new ShallowRenderer();
  shallowRenderer.render(component);
  const result = shallowRenderer.getRenderOutput();

  return result;
};

export const shouldShallowMatchSnapshot = (title, component) => {
  it(title, () => {
    const tree = shallowRender(component);
    expect(tree).toMatchSnapshot();
  });
};

export const isNull = (title, component) => {
  it(title, () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toBeNull();
  });
};

const errorIfMissingKey = (keys, object, message) => {
  keys.forEach(key => {
    if (!(key in object)) {
      throw new Error(`Missing value '${key}' in ${message}.`);
    }
  });
};

const checkKeysExistInBothObjects = (object1, object2, message1, message2) => {
  const object1Keys = Object.keys(object1);
  const object2Keys = Object.keys(object2);

  errorIfMissingKey(object1Keys, object2, message2);
  errorIfMissingKey(object2Keys, object1, message1);
};

const checkTypesOfExports = (
  actualExportsByName,
  actualExports,
  expectedExports,
  utilityName,
) => {
  actualExportsByName.forEach(actualExportName => {
    const actualExportValue = actualExports[utilityName][actualExportName];
    const expectedExport = expectedExports[utilityName][actualExportName];
    const typeCheck = typeof actualExportValue === expectedExport; // eslint-disable-line valid-typeof

    // if this fails it is likely that an export is missing from the unit test expectation
    expect(typeCheck).toBe(true);
  });
};

export const testUtilityPackages = (
  actualExports,
  expectedExports,
  packageName,
) => {
  const actualUtilities = Object.keys(actualExports);

  // check if the actual exported file has defined expected values to match and that the expected values are actually exported
  checkKeysExistInBothObjects(
    actualExports,
    expectedExports,
    `the actual utilities for '${packageName}'`,
    `the expected utilities for '${packageName}'`,
  );

  actualUtilities.forEach(utilityName => {
    const actualExportsByName = Object.keys(actualExports[utilityName]);

    // check if each of the actual exported consts have an expected value to match and that the expected values are actually exported
    checkKeysExistInBothObjects(
      actualExports[utilityName],
      expectedExports[utilityName],
      `the actual export for '${packageName}/${utilityName}'`,
      `the expected export for '${packageName}/${utilityName}'`,
    );

    checkTypesOfExports(
      actualExportsByName,
      actualExports,
      expectedExports,
      utilityName,
    );
  });
};
