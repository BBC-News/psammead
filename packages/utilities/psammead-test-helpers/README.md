# psammead-test-helpers &middot; [![GitHub license](https://img.shields.io/badge/license-Apache%202.0-blue.svg)](https://github.com/BBC-News/psammead/blob/latest/LICENSE) [![npm version](https://img.shields.io/npm/v/@bbc/psammead-test-helpers.svg)](https://www.npmjs.com/package/@bbc/psammead-test-helpers) [![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg)](https://github.com/BBC-News/psammead/blob/latest/CONTRIBUTING.md)

This package is a collection helper methods used for snapshot testing, with support for styled components.

## Installation

```jsx
npm install react react-dom @bbc/psammead-test-helpers --save-dev
```

## Usage

```jsx
import { shouldMatchSnapshot } from '@bbc/psammead-test-helpers';

shouldMatchSnapshot(
  'should render correctly',
  <h1>Hello World</h1>,
);
```

## Functions

| Name                       | Arguments        | Description  |
|:---------------------------|:-----------------|:-------------|
| shouldMatchSnapshot        | title, component | Renders the component using react-test-renderer, converts it to JSON and asserts that it matches the given snapshot, which will be saved in the `__snapshots__` directory. The first argument `title` is the title for the test. |
| shallowRender              | component        |  Shallow renders the component using react-test-renderer and returns the render output |
| shouldShallowMatchSnapshot | title, component | Shallow renders the component using react-test-renderer and asserts that it matches the given snapshot, which will be saved in the `__snapshots__` directory. The first argument `title` is the title for the test. |
| isNull                     | title, component    | Renders the component using react-test-renderer, converts it to JSON and asserts that it is null. The first argument `title` is the title for the test. |

## Roadmap

There is currently a plan to migrate from `react-test-renderer` to `enzyme` as it provides a nicer higher-level API over the same functionality, as well as having support for more granular unit tests.

## Contributing

Psammead is completely open source. We are grateful for any contributions, whether they be new components, bug fixes or general improvements. Please see our primary contributing guide which can be found at [the root of the Psammead respository](https://github.com/BBC-News/psammead/blob/latest/CONTRIBUTING.md).

### [Code of Conduct](https://github.com/BBC-News/psammead/blob/latest/CODE_OF_CONDUCT.md)

We welcome feedback and help on this work. By participating in this project, you agree to abide by the [code of conduct](https://github.com/BBC-News/psammead/blob/latest/CODE_OF_CONDUCT.md). Please take a moment to read it.

### License

Psammead is [Apache 2.0 licensed](https://github.com/BBC-News/psammead/blob/latest/LICENSE).
