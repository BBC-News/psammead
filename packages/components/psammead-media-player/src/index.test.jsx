import React from 'react';
import { shouldMatchSnapshot } from '@bbc/psammead-test-helpers';
import { CanonicalMediaPlayer, AmpMediaPlayer } from '.';

describe('Media Player: AMP Entry', () => {
  shouldMatchSnapshot(
    'renders a landscape container with an amp-iframe and nested amp-img',
    <AmpMediaPlayer
      placeholderSrc="http://foo.bar/placeholder.png"
      src="http://foo.bar/iframe/amp"
      title="Media player"
      height={16}
      width={9}
    />,
  );

  shouldMatchSnapshot(
    'renders a portrait container with amp-iframe and nested amp-img',
    <AmpMediaPlayer
      portrait
      placeholderSrc="http://foo.bar/placeholder.png"
      src="http://foo.bar/iframe/amp"
      title="Media player"
      height={9}
      width={16}
    />,
  );

  shouldMatchSnapshot(
    'renders the audio skin',
    <AmpMediaPlayer
      showPlaceholder={false}
      src="https://www.test.bbc.com/ws/av-embeds/media/bbc_korean_radio/liveradio"
      skin="audio"
      placeholderSrc="http://foo.bar/placeholder.png"
      title="Audio player"
      height={9}
      width={16}
    />,
  );
});

describe('Media Player: Canonical Entry', () => {
  const mediaInfo = {
    duration: '2:30',
    durationSpoken: '2 minutes 30 seconds',
    datetime: 'PT2M30S',
  };

  shouldMatchSnapshot(
    'renders a landscape container with a placeholder image',
    <CanonicalMediaPlayer
      placeholderSrc="http://foo.bar/placeholder.png"
      src="http://foo.bar/iframe"
      service="news"
      mediaInfo={{ title: 'Dog chases cat.', ...mediaInfo }}
      title="Media player"
    />,
  );

  shouldMatchSnapshot(
    'renders a portrait container with a placeholder image',
    <CanonicalMediaPlayer
      placeholderSrc="http://foo.bar/placeholder.png"
      src="http://foo.bar/iframe"
      portrait
      service="news"
      mediaInfo={{ title: 'Dog chases cat.', ...mediaInfo }}
      title="Media player"
    />,
  );

  shouldMatchSnapshot(
    'renders an iframe when showPlaceholder is false',
    <CanonicalMediaPlayer
      showPlaceholder={false}
      src="http://foo.bar/iframe"
      service="news"
      mediaInfo={{ title: 'Dog chases cat.', ...mediaInfo }}
      title="Media player"
    />,
  );

  shouldMatchSnapshot(
    'renders the audio skin',
    <CanonicalMediaPlayer
      showPlaceholder={false}
      src="https://www.test.bbc.com/ws/av-embeds/media/bbc_korean_radio/liveradio"
      skin="audio"
      service="news"
      mediaInfo={{ type: 'audio', title: 'Dog barks at cat.', ...mediaInfo }}
      title="Audio player"
    />,
  );
});
