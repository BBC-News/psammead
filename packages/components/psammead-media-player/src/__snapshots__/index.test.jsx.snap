// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Media Player: AMP Entry renders a landscape container with an amp-iframe and nested amp-img 1`] = `
.c0 {
  padding-top: 56.25%;
  position: relative;
  overflow: hidden;
}

<html>
  <head>
    <script
      async="true"
      custom-element="amp-iframe"
      src="https://cdn.ampproject.org/v0/amp-iframe-0.1.js"
    />
  </head>
  <body>
    <div>
      <div
        class="c0"
      >
        <amp-iframe
          allowfullscreen="allowfullscreen"
          frameborder="0"
          layout="fill"
          sandbox="allow-scripts allow-same-origin"
          src="http://foo.bar/iframe/amp"
          title="Media player"
        >
          <amp-img
            attribution=""
            layout="fill"
            placeholder="true"
            src="http://foo.bar/placeholder.png"
          />
        </amp-iframe>
      </div>
    </div>
  </body>
</html>
`;

exports[`Media Player: AMP Entry renders a portrait container with amp-iframe and nested amp-img 1`] = `
<html>
  <head>
    <script
      async="true"
      custom-element="amp-iframe"
      src="https://cdn.ampproject.org/v0/amp-iframe-0.1.js"
    />
  </head>
  <body>
    .c0 {
  padding-top: 177.78%;
  position: relative;
  overflow: hidden;
}

<div>
      <div
        class="c0"
      >
        <amp-iframe
          allowfullscreen="allowfullscreen"
          frameborder="0"
          layout="fill"
          sandbox="allow-scripts allow-same-origin"
          src="http://foo.bar/iframe/amp"
          title="Media player"
        >
          <amp-img
            attribution=""
            layout="fill"
            placeholder="true"
            src="http://foo.bar/placeholder.png"
          />
        </amp-iframe>
      </div>
    </div>
  </body>
</html>
`;

exports[`Media Player: AMP Entry renders the audio skin 1`] = `
<html>
  <head>
    <script
      async="true"
      custom-element="amp-iframe"
      src="https://cdn.ampproject.org/v0/amp-iframe-0.1.js"
    />
  </head>
  <body>
    .c0 {
  height: 165px;
  position: relative;
  overflow: hidden;
}

<div>
      <div
        class="c0"
      >
        <amp-iframe
          allowfullscreen="allowfullscreen"
          frameborder="0"
          layout="fill"
          sandbox="allow-scripts allow-same-origin"
          src="https://www.test.bbc.com/ws/av-embeds/media/bbc_korean_radio/liveradio"
          title="Audio player"
        >
          <amp-img
            attribution=""
            layout="fill"
            placeholder="true"
            src="http://foo.bar/placeholder.png"
          />
        </amp-iframe>
      </div>
    </div>
  </body>
</html>
`;

exports[`Media Player: Canonical Entry renders a landscape container with a placeholder image 1`] = `
.c9 {
  display: block;
  width: 100%;
  visibility: visible;
}

.c7 {
  vertical-align: middle;
  margin: 0 0.25rem;
  color: #222222;
  fill: currentColor;
  width: 0.75rem;
  height: 0.75rem;
}

.c5 {
  -webkit-clip-path: inset(100%);
  clip-path: inset(100%);
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  height: 1px;
  overflow: hidden;
  position: absolute;
  width: 1px;
  margin: 0;
}

.c3 {
  background-color: #222222;
  border: none;
  color: #FFFFFF;
  cursor: pointer;
  display: block;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-weight: 700;
  font-style: normal;
  font-size: 0.75rem;
  line-height: 1rem;
  height: 5rem;
  padding: 0;
  -webkit-transition: background-color 300ms;
  transition: background-color 300ms;
  width: 5rem;
}

.c3:hover,
.c3:focus {
  background-color: #B80000;
  -webkit-transition: background-color 300ms;
  transition: background-color 300ms;
}

.c6 > svg {
  color: #FFFFFF;
  fill: currentColor;
  height: 1.5rem;
  margin-top: 0.5rem;
  width: 1.5rem;
}

.c8 {
  display: block;
  margin-top: 0.5rem;
}

.c2 {
  cursor: pointer;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
}

.c4 {
  position: absolute;
  bottom: 0;
}

.c1:hover .c4,
.c1:focus .c4 {
  background-color: #B80000;
}

.c0 {
  padding-top: 56.25%;
  position: relative;
  overflow: hidden;
}

<div
  class="c0"
>
  <div
    class="c1 c2"
  >
    <button
      class="c3 c4"
    >
      <span
        class="c5"
      >
        Play video, "Dog chases cat.", 2 minutes 30 seconds
      </span>
      <div
        aria-hidden="true"
        class="c6"
      >
        <svg
          class="c7"
          focusable="false"
          height="12px"
          viewBox="0 0 32 32"
          width="12px"
        >
          <polygon
            points="3,32 29,16 3,0"
          />
        </svg>
      </div>
      <time
        aria-hidden="true"
        class="c8"
        datetime="PT2M30S"
      >
        2:30
      </time>
    </button>
    <img
      alt=""
      class="c9"
      src="http://foo.bar/placeholder.png"
    />
  </div>
</div>
`;

exports[`Media Player: Canonical Entry renders a portrait container with a placeholder image 1`] = `
.c9 {
  display: block;
  width: 100%;
  visibility: visible;
}

.c7 {
  vertical-align: middle;
  margin: 0 0.25rem;
  color: #222222;
  fill: currentColor;
  width: 0.75rem;
  height: 0.75rem;
}

.c5 {
  -webkit-clip-path: inset(100%);
  clip-path: inset(100%);
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  height: 1px;
  overflow: hidden;
  position: absolute;
  width: 1px;
  margin: 0;
}

.c3 {
  background-color: #222222;
  border: none;
  color: #FFFFFF;
  cursor: pointer;
  display: block;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-weight: 700;
  font-style: normal;
  font-size: 0.75rem;
  line-height: 1rem;
  height: 5rem;
  padding: 0;
  -webkit-transition: background-color 300ms;
  transition: background-color 300ms;
  width: 5rem;
}

.c3:hover,
.c3:focus {
  background-color: #B80000;
  -webkit-transition: background-color 300ms;
  transition: background-color 300ms;
}

.c6 > svg {
  color: #FFFFFF;
  fill: currentColor;
  height: 1.5rem;
  margin-top: 0.5rem;
  width: 1.5rem;
}

.c8 {
  display: block;
  margin-top: 0.5rem;
}

.c2 {
  cursor: pointer;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
}

.c4 {
  position: absolute;
  bottom: 0;
}

.c1:hover .c4,
.c1:focus .c4 {
  background-color: #B80000;
}

.c0 {
  padding-top: 177.78%;
  position: relative;
  overflow: hidden;
}

<div
  class="c0"
>
  <div
    class="c1 c2"
  >
    <button
      class="c3 c4"
    >
      <span
        class="c5"
      >
        Play video, "Dog chases cat.", 2 minutes 30 seconds
      </span>
      <div
        aria-hidden="true"
        class="c6"
      >
        <svg
          class="c7"
          focusable="false"
          height="12px"
          viewBox="0 0 32 32"
          width="12px"
        >
          <polygon
            points="3,32 29,16 3,0"
          />
        </svg>
      </div>
      <time
        aria-hidden="true"
        class="c8"
        datetime="PT2M30S"
      >
        2:30
      </time>
    </button>
    <img
      alt=""
      class="c9"
      src="http://foo.bar/placeholder.png"
    />
  </div>
</div>
`;

exports[`Media Player: Canonical Entry renders an iframe when showPlaceholder is false 1`] = `
.c1 {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  border: 0;
  overflow: hidden;
}

.c0 {
  padding-top: 56.25%;
  position: relative;
  overflow: hidden;
}

<div
  class="c0"
>
  <iframe
    allow="autoplay; fullscreen"
    allowfullscreen=""
    class="c1"
    src="http://foo.bar/iframe"
    title="Media player"
  />
</div>
`;

exports[`Media Player: Canonical Entry renders the audio skin 1`] = `
.c1 {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  border: 0;
  overflow: hidden;
}

.c0 {
  height: 165px;
  position: relative;
  overflow: hidden;
}

<div
  class="c0"
>
  <iframe
    allow="autoplay; fullscreen"
    allowfullscreen=""
    class="c1"
    src="https://www.test.bbc.com/ws/av-embeds/media/bbc_korean_radio/liveradio"
    title="Audio player"
  />
</div>
`;
